### ISSUE TITLE
Proposal: Support function parameter type constraints/wildcards

### ISSUE THREAD

**Issue by topolarity** [ziglang/zig#9723]
### Motivation

The goals of this proposal are to:
  1. Make generic function prototypes more self-documenting 
  2. Encourage the enforcement of simple type constraints at compile-time
  3. Make the use of runtime and `comptime` polymorphism more symmetric, in terms of safety and syntax

The key mechanism I propose is a limited, predictable form of pattern matching for types in function parameters, which is intended to allow straightforward enforcement of type constraints at compile-time and run-time.

### Part A: Introduce `any T` as a type constructor wildcard

Pattern matching at compile-time amounts to constrained `comptime` polymorphism. This can make it easier to infer function behavior from its prototype alone.

Compare the monomorphic version of a basic `append` function versus its generic counterparts:

```zig
// Monomorphic
fn       u8_append(l: ArrayList(u8), x: u8) // <- Fully-monorphized, clear types

// Compile-time Polymorphic
fn generic_append1(l: anytype, x: anytype)  // <- Option 1: Have to guess or read implementation
fn generic_append2(comptime T: type, l: *ArrayList(T), x: T)  // <- Option 2: Better, but this is _less_ convenient
                                                              //    at the call-site, discouraging use!
```

These functions are intended to generalize `u8_append` by allowing variance over the inner type parameter `u8`. However, the programmer must make a difficult choice: Either they write `generic_append1`, providing an uninformative function signature which support convenient type inference, or they write `generic_append2` with an informative function signature and automatically-enforced constraints, but which requires explicitly passing in the correct type separately.

Wouldn't it be nice if we could just express this variance directly? We'd ideally like to be able to write functions like:

```zig
fn append(l: *ArrayList(any), x: any) void      // More readable
fn append(l: *ArrayList(any T), x: any T) void  // Even better, fully constrained!
```

As you can see, by allowing implicit equality constraints between wildcards, this also provides a simple declarative syntax for expressing certain parameter type requirements, such as when two types must agree, and allows expressing some parameter-dependent return types more directly (avoiding complex `@TypeInfo(TypeOf(param))` expressions in some cases).

```zig
fn foo(a: any T, b: any T, c: any U, d: any U) Tuple(.{T,U})  // a & b must share type, c & d must share type
```

#### Changes required

At first glance, this inference appears to be a basic unification problem. However, the proposal above is incompatible with Zig’s existing `fn(…) type` type constructors, since multiple functions may return instances of the same `struct {…}` definition and also perform non-invertible transformations on their parameters.

So naively adding pattern matching like this for `fn(…) type` is undecidable for the compiler, and even if it weren’t, it would be confusing and unpredictable to the user, given multiple overlapping “interfaces” to one underlying type.

The problem here is that `type` objects have no canonical constructor function, whose parameters we could associate with the `type` and use for unification. To enable this, we take inspiration from #1717 and borrow the `fn` syntax for `struct`.

A `struct` block can now take its own set of parameters:

```zig
const FixedArray = struct(T: type, N: usize) {
  contents: [N]T
};
var x: FixedArray(u8, 5) =  ... 
```

`FixedArray` is called a **type constructor function** (a `type(...)`), distinguishing these from fully-constructed types (e.g. `FixedArray(u8, 5)` or `u8`) and existing function-returning types (`fn(…) type`).

The key restriction that makes this proposal possible is that **we only allow inferring parameters to `type(...)` functions**, not to `fn(...) type` functions. In the compiler, a constructed `type` saves the parameters provided to its `type(...)` function, so that these can be used for inference at the call site to a function as above.

In general, a `type(A,B,C)` can be used like a `fn(A,B,C) type`, except that it supports type inference. With this restriction, the inference problem is decidable for the compiler - it becomes “word matching”, with the input type containing no metavariables. It’s also intended to be easy to understand for the user (feedback welcome, of course).

#### Extending to `union`

For completeness, `union` will need the same treatment as `struct`, but we already have a `union(tag)` syntax for tagged unions.

Thankfully, the existing syntax isn't incompatible with a naive solution, although it is a bit awkward:

```zig
// Tagged
const Result = union(enum)(T: type) {
    ok: T,
    not_ok: void,
};
// Untagged
const SpecialEnum = union(T: type) {
    default: T,
    int: u32,
    float: f32,
};
```

Any suggestions to improve this syntax are more than welcome.

#### Backwards compatibility

Existing usage of `fn(...) type` would be backwards-compatible with this change, but these `type` objects are not `type(...)` functions, so they cannot be used for inference. To support inference, the module author has to expose a `type(...)` function, either by defining it at the top-level or, less commonly, returning one from a function unevaluated.

### Part B: Runtime-Erasure via `erased T`

As an extension to this proposal, we can support `erased T` to indicate an inferred parameter that should be constrained but not used to specialize the function. Note that this is the typical pattern used for **runtime polymorphism**.

To put some familiar examples in the new terminology:
 - C’s use of `void *` as a run-time polymorphic type is really a `* erased T`
 - Arrays of unknown size, instead of being written `[*]T`, can be viewed as `*[erased]T`.

In each of these cases, the types pass through a function which “forgets” a certain part of the concrete type it operates on, and then delegates to other functions implemented on the concrete (fully-specified) type.

#### Mechanics

When `erased T` is used as a wildcard pattern for a function parameter, any constraints on `T` are still enforced statically at the call-site, as usual. However, `T` is not allowed to be used inside the function, and the function is _not_ compile-time specialized (i.e. monomorphized) on `T`.

This feature is useful, for example, for ensuring at compile-time that multiple parameters are type-compatible, without specializing a function at `comptime`:

```zig
fn accumulate(a: *LinkedList(erased T), b: *fn(*erased T) u64) u64
```

This can also be used to make sure that a method dispatch table (i.e. vtable) is compatible with an object it is used with. It also makes the use of run-time vs. `comptime` polymorphism more symmetric.

Here’s an example, where the “vtable” is a single function pointer:

```zig
// run-time polymorphic
fn accumulate(a: *LinkedList(erased T), b: *fn(*erased T) u64) u64
// comptime polymorphic
fn accumulate(a: *LinkedList(any T), comptime b: fn(any T) u64) u64
```

As you’d expect with run-time polymorphic interfaces, `erased` types live behind pointers since their sizes are unknown. Such a pointer must be cast to a non-erased `type` before being dereferenced.

**Note:** In theory, it may be possible to dereference and use an `erased` type for a limited set of accesses/operations, provided that the operation is solely a function of its non-erased parameters. However, such behavior is considered **out-of-scope** for this proposal.

#### Debug Safety

I believe this opens the door to additional run-time safety in debug builds: For instrumentation, the compiler should be able to replace an `erased` pointer with a fat pointer describing its `erased` parameters. Then, when a specialized function `fn(x: *A)` is cast to an `erased` function pointer `*fn(x: * erased T)`, it can be wrapped in a check that inspects the fat pointer information to make sure that the underlying type is `A`.

### Appendix: Type constructor functions as variables ("Sandwich" Inference)

"Sandwich" inference is when one of the type constructor functions in the signature is a metavariable to be inferred.

The sandwich part refers to the fact that the variables can appear in-between constant terms, rather than just at the top-most level of parameters. This can lead to pretty strange and complex patterns:

```zig
fn(x: any L(any T, any N), x: T[N]) u32
fn(x: any T(u32, u8)) u32 
fn(x: ArrayList(any T(u32, u8))) u32 
```

This inference is perfectly decidable as a unification problem, but these patterns often feel more confusing than helpful. For that reason, it may be preferable to **forbid** sandwich inference, at least until a clear use case arises.

### Appendix: Syntax and keywords

In order to disambiguate between inferred parts of a term versus constants, we should require `erased`/`any` for every appearance of a metavariable `T`. Of course, a metavariable `T` should never shadow a constant.

The remaining details of the syntax are much more debatable, in particular the usage of `erased`/`any`. There are other variations we might consider, as well:
```zig
                       Option 1       Option 2        Option 3     Option 4
Comptime specialized:      any T | comptime any T |       any T |    var T
Run-time      erased:   erased T |          any T | undefined T | erased T
```

I've been preferring Option 1 on the grounds that Option 2 is too verbose and Option 4 is misleading by labelling comptime specialization with the `var` keyword. Option 3 is a good candidate as well, since it doesn't require introducing a new keyword, but it seems unintuitive to have a potentially-constrained metavariable labelled with `undefined`, even if it's true that you're not allowed to use it within the function.

Of course, this discussion is closely related to #5893

If this proposal is accepted, we might also want to consider an alternative slice syntax such as `*[var]T`, since many newcomers to the language seem to be suprised that `[]T` is actually a reference type.

### Appendix: @TypeInfo/@Type

`@TypeInfo`/`@Type` would be supported exactly as they are today. In particular, they would _not_ support creating a `type(...)` function.

There are places in the standard library (e.g. `enums.zig`, `mem.zig`, `meta.zig`) that use `@Type` today. These use cases would benefit from a built-in `@TypeFunction()` which converts a `fn(...) TypeInfo` function into a `type(...)` function, if inference is desired for these families of reified types.

**Comment by zzyxyzz** [comment]
Over in #9260, there was a lot of discussion about how to make something like `ArrayList(any T)` work within the current language semantics. The best insight was that, with comptime functions being memoized and structs being named after the function they are [returned from](https://ziglang.org/documentation/master/#Struct-Naming), the job is *almost* done for us. An `ArrayList(i32)` should have exactly that as its `@typeName`, and matching it against `ArrayList(any T)` should be trivial. Except that `ArrayList(T)` is implemented as an alias of `ArrayListAligned(T, null)`, which is its *actual* type name, and matching that against `ArrayList(T)` remains a hard problem in the general case. So far nobody seems to have figured out a complete solution to this.

Introducing proper parametric types is a possible way out, but then we have two different systems for generic types: One can still return anonymous types from comptime functions, which gives introspection, reification and custom `@compileError`s, but no inference. Or one can use the new parametric types, which support inference and type constraints, but no complex logic. We end up with two mutually incompatible solutions that each solve half of the problem. That is not a good place to be in, IMHO.

**Comment by topolarity** [comment]
> Over in #9260, there was a lot of discussion about how to make something like `ArrayList(any T)` work within the current language semantics.

Thanks for the reference! I hadn't seen that issue yet. Reading through it was informative for the challenges of supporting this kind of inference without explicitly separating type constructor functions syntactically.

One "automatic" equivalent for this proposal would be to enable support type-inference only for the inner-most function including a `@Type` or `struct { ... }` expression. This would be functionally the same without needing to introduce `struct(...)`. I left this alternative out of the proposal, thinking it was too implicit/incomplete to be intuitive, but perhaps it's worth considering

> Introducing proper parametric types is a possible way out, but then we have two different systems for generic types: One can still return anonymous types from comptime functions, which gives introspection, reification and custom `@compileError`s, but no inference. Or one can use the new parametric types, which support inference and type constraints, but no complex logic. We end up with two mutually incompatible solutions that each solve half of the problem. That is not a good place to be in, IMHO.

I definitely agree that adding `type(A,B,C)` as an alternative to `fn(A,B,C) type` is a syntactic/conceptual overhead that comes with a significant cost. 

On the other hand, I'm not sure that I agree that there's a feature gap between `fn(...) type` and `type(...)`, outside of parameter inference. Fully-concrete types are still just `type`, supporting introspection, reification, etc. Meanwhile, parametric families of types (`type(A,B,C)` and `fn(A,B,C) type`) do not support introspection, but do support reification, at least if we include `@TypeFunction`. Can you spot a counter-example of something that's supported by `fn(...) type` and not `type(...)`?

**Comment by zzyxyzz** [comment]
> Can you spot a counter-example of something that's supported by fn(...) type and not type(...)?

I was thinking along the lines of additional constraints and comptime checks on the type parameter. But you're right, those can be implemented with helper functions, or even inline:
```zig
const Foo = struct(T: type) {
    value: if (meta.trait.isNumber(T)) T else @compileError("Not a number!"),
};
```

One thing you can't do with parametric structs is add or omit fields depending on the inputs, but I don't think this is used much anyway. There might be more, but I can't think of anything right now.

---

I'm now sort of in favor of this proposal. The idea to build a generic type system from first-class types and compile-time code execution is both clever and powerful, but it creates a lot of dark corners semantically, and makes many things that work very simply in other languages either difficult or impossible.  From a usability perspective, having a simple parametric type system, and maybe interfaces, would remove all magic from 95% of the use cases. Personally, I think that would be an appropriate power-complexity tradeoff for a language that aims to be a C replacement. But this is also a fairly significant change, which may have ripple effects throughout the type system. I'm not sure whether it isn't too late for that.

**Comment by topolarity** [comment]
Since there are quite a few concepts in play now, here's a quick glossary in plain English.

Any "hole" in a type pattern has 3 different "ABI modes" that affect casting and specialization:
- `any` -> "This function needs to read/discriminate this value/type. Please specialize the function on this value/type"
- `var` -> "This function needs to read/discriminate this value/type. However, please **don't** specialize it on this parameter."
- `erased` -> "This function does not need to read/discriminate this value/type. However, please add a debug check to make sure I pass it around correctly to functions who _do_ care."

In some sense, `var`'s behavior is dual to `comptime`. It takes a typically-comptime value and lifts it to run-time, while `comptime` takes what would be a run-time value and lowers it to comptime. 

Meanwhile, `erased` is an optimization for when we actually don't need to use the parameter at all.

This also gives some intuition for why many important C foot guns (including handling structs with flexible array members) can be seen as casting to a partially-`erased` type. These are very dangerous in C since there are no special debug-mode checks. "Fat" pointers meanwhile are a manual wrapper around such an `erased` type, and so are un-ergonomic and easy to accidentally separate from the value whose types they describe
